include/dlaf/common/unwrap.h

```diff
@@ -58,7 +59,17 @@ struct Unwrapping {
   template <typename... Ts>
   auto operator()(Ts&&... ts)
       -> decltype(std::move(f)(Unwrapper<std::decay_t<Ts>>::unwrap(std::forward<Ts>(ts))...)) {
-    return std::move(f)(Unwrapper<std::decay_t<Ts>>::unwrap(std::forward<Ts>(ts))...);
+    using result_type =
+        decltype(std::move(f)(Unwrapper<std::decay_t<Ts>>::unwrap(std::forward<Ts>(ts))...));
+    if constexpr (std::is_void_v<result_type>) {
+      std::move(f)(Unwrapper<std::decay_t<Ts>>::unwrap(std::forward<Ts>(ts))...);
+      std::tuple<Ts...>(std::forward<Ts>(ts)...);
+    }
+    else {
+      auto r = std::move(f)(Unwrapper<std::decay_t<Ts>>::unwrap(std::forward<Ts>(ts))...);
+      std::tuple<Ts...>(std::forward<Ts>(ts)...);
+      return r;
+    }
```

Comment:
I'd like to see this separated out into a separate function wrapper. Perhaps called `consume_rvalues` or similar (but we can bikeshed the name separately). The important thing is to separate independent functionality into independent function wrappers. I think you can mostly follow the pattern of unwrap except that you can simplify it significantly since you don't need customization for the consuming rvalues functionality.

Additionally, I'd like to see benchmarks of this change before it's merged. I wouldn't expect a big difference. I'd be sad and slightly surprised if performance is worse, but don't rule it out. Ideally we'd see a small, but measurable performance improvement from this.